[Hub Generated] Review request for Microsoft.SqlVirtualMachine to add version preview/2022-08-01-preview

qianliu-pdx opened the pull request:
This is a PR generated at [OpenAPI Hub](https://aka.ms/openapihub). You can view your [work branch via this link](https://portal.azure-devex-tools.com/branch/qianliu-pdx/azure-rest-api-specs/dev-sqlvirtualmachine-Microsoft.SqlVirtualMachine-2022-08-01-preview...main/).
## ARM API Information (Control Plane)


Azure 1st Party Service can try out the [Shift Left](https://aka.ms/ShiftLeft) experience to initiate API design review from ADO code repo.  If you are interested, may request engineering support by filling in with the form https://aka.ms/ShiftLeftSupportForm.

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [x] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. By default, Azure SDKs of all languages (.NET/Python/Java/JavaScript for both management-plane SDK and data-plane SDK, Go for management-plane SDK only ) MUST be refreshed with/after swagger of new version is published. If you prefer NOT to refresh any specific SDK language upon swagger updates in the current PR, please leave details with justification here.

### Contribution checklist (MS Employees Only):
- [x] I commit to follow the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist

> **Applicability**: :warning:
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs
> - Adding new properties
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [x] Check this box if any of the following apply to the PR so that the label "ARMReview" and "WaitForARMFeedback" will be added by bot to kick off ARM API Review.  Missing to check this box in the following scenario may result in delays to the ARM manifest review and deployment.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit and then push new changes, including version updates, in separate commits. You can use OpenAPIHub to initialize the PR for adding a new version. For more details refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

- [x] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them.

### Breaking Change Review Checklist
If you have any breaking changes as defined in the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy/), request approval from the Breaking Change Review Board.

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Additional details on the process and office hours are on the [Breaking Change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

NOTE: To update API(s) in public preview for over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

qianliu-pdx committed:
Adds base for updating Microsoft.SqlVirtualMachine from version preview/2022-07-01-preview to version 2022-08-01-preview

qianliu-pdx committed:
Updates readme

qianliu-pdx committed:
Updates API version in new specs and examples

Commit:
Adding 2022-08-01-preview version for Sql Virtual Machine swagger

Commit:
move all api-version to 2022-08-01-preview

Commit:
fix the following validation errors:

R4041 - XmsIdentifierValidation
R4009 - RequiredReadOnlySystemData

qianliu-pdx commented:
/azp run unifiedPipeline

Commit:
add location for 202 reponse

Commit:
fix typo

qianliu-pdx commented:
/azp run unifiedPipeline

Commit:
fix LRO_RESPONSE_HEADER validation error

qianliu-pdx commented:
/azp run unifiedPipeline

Commit:
fix Swagger PrettierCheck

rchlkm commented on specification/sqlvirtualmachine/resource-manager/Microsoft.SqlVirtualMachine/preview/2022-08-01-preview/sqlvm.json:
```diff
@@ -0,0 +1,2542 @@
+          ],
+          "type": "string",
+          "x-ms-enum": {
+            "name": "IdentityTypeWithNone",
```
Please manually change the value to "IdentityType".
This is a known issue mentioned in one of our team's wikis.
This will resolve the 1047 - XmsEnumChanged Breaking change validation error

rchlkm commented on specification/sqlvirtualmachine/resource-manager/Microsoft.SqlVirtualMachine/preview/2022-08-01-preview/sqlvm.json:
```diff
@@ -0,0 +1,2542 @@
+          "$ref": "#/definitions/SQLInstanceSettings",
+          "description": "SQL Instance settings."
+        },
+        "azureADAuthenticationSettings": {
```
seems like this is violating the camel case lint rule (DefinitionsPropertiesNamesCamelCase).
We would need to update it as "azureAdAuthenticationSettings" in the RP code as well

qianliu-pdx replied:
followed your suggestion to add azureADAuthenticationSettings to custom-words.txt

rchlkm replied:
@akning-ms Could you give us suggestion on how to fix the LintDiff DefinitionsPropertiesNamesCamelCase error? Since azureAD is the name of a service, we decied to add it to the custom-words list but the validations are still throwing the error.

leni-msft commented:
/azp run

ruowan commented:
/azp run

ruowan commented:
/azp run

azure-sdk commented:
**API change check**

APIView has identified API level changes in this PR and created following API reviews.

[track2_azure-mgmt-sqlvirtualmachine](https://apiview.dev/Assemblies/Review/1fece2a8ffc94131a60106c0549f0f9a)
[sdk/resourcemanager/sqlvirtualmachine/armsqlvirtualmachine](https://apiview.dev/Assemblies/Review/67ba43756c894146add3413461b5d573)
[Microsoft.SqlVirtualMachine](https://apiview.dev/Assemblies/Review/e660be3521ce4148bd7aa705cac49ded)

Commit:
fix 1047 - XmsEnumChanged, manually move IdentityTypeWithNone to
IdentityType

Commit:
fix DefinitionsPropertiesNamesCamelCase, adding
azureADAuthenticationSettings to the custom-words.txt

Commit:
adding azureADAuthenticationSettings in the alphabetical position.

Commit:
fix PostOperationAsyncResponseValidation and  LroLocationHeader errors

Commit:
fix typo

Commit:
missing quote

Commit:
adding quote

jianyexi commented:
For the linting error, you can rename the property to came case 'azureAdAuthenticationSettings' to solve it

Commit:
fix DefinitionsPropertiesNamesCamelCase

Commit:
remove azureADAuthenticationSettings

Commit:
fix ResourceNameRestriction

Commit:
use \" to escape "

Commit:
fix swagger regex syntax difference comparing to C# regex

qianliu-pdx commented:
@leni-msft @jianyexi , there is no validation error marked as "Required", could you please review this PR? Thanks.

rkmanda commented:
>             "description": "*** Error Responses: ***\n\n * 400 InvalidParameterValue - An invalid value was given to parameter.\n\n * 400 RPGenericUser - User Failure when calling other Resource Provider.\n\n * 400 RPResponseMissingAsyncOperationHeader - Response is missing Async operation header information\n\n * 400 InvalidArgument - Invalid argument '{0}'.\n\n * 400 ArgumentNotInRange - Argument '{0}' not in range.\n\n * 400 ResourceNotProvisioned - As the resource: {0} is not in a provisioned state, the request cannot be proceeded forward\n\n * 400 CRPNotAllowedOperation - Operation cannot be completed due to the following error: {0}\n\n * 400 AadAppDisabled - Microsoft AAD app SQLVMResourceProviderAuth ({0}) is disabled in your tenant.\n\n * 400 InternalAPIAccessDenied - Internal API Access denied. App Id {0} is not valid to use this API.\n\n * 400 InvalidSQLAuthPasswordLength - SQL auth password should not be greater than 128 characters\n\n * 400 InvalidRgResourceId - Invalid Resourcegroup resource id specified.\n\n * 400 VmNicVnetMismatch - Virtual machine NIC VNet mismatch.\n\n * 400 NoAvailabilitySet - Vm is not associated with any availability set.\n\n * 400 AvailabilitySetMismatch - Availability set of virtual machines does not match.\n\n * 400 MismatchVmGroupSubscription - Subscription id for SQL virtual machine and SQL virtual machine group are different.\n\n * 400 VmNotRunning - The VM is not in running state.\n\n * 400 VmAgentNotRunning - The VM agent is not installed or in running state.\n\n * 400 AgDoesNotExist - Availability group does not exist.\n\n * 400 AgListDoesNotMatch - SQL Availability group list does not match the list of virtual machines in SQL Virtual machine group.\n\n * 400 NoIpProvided - At least one IP needs to be provided.\n\n * 400 MoreIpProvided - Only one IP needs to be provided.\n\n * 400 SqlVmGroupDoesNotHaveAGListener - SQL virtual machine group does not have the AG listener.\n\n * 400 AgListenerUpdateNotAllowed - Only handful properties of availability group listener can be updated.\n\n * 400 SqlVmNotInGroup - SQL virtual machine is not part of the group.\n\n * 400 SqlVmListCannotBeEmpty - SQL virtual machines list cannot be empty.\n\n * 400 PublicIpNotIPv4 - Public IP must be IPv4 address.\n\n * 400 PublicIPDynamicAllocation - Public IP allocation mode must be static\n\n * 400 PublicLBInvalid - Load balancer specified is not public.\n\n * 400 SubnetMissingFromPrivateIP - Subnet must be provided with private IP.\n\n * 400 IPNotInSubnet - IP not part of subnet.\n\n * 400 NoActivePrimaryInAG - None of the nodes in AvailabilityGroup are Primary\n\n * 400 MultipleListenerSameAG - Multiple availability group listeners for same availability group are not allowed.\n\n * 400 AgListenerNotEmpty - Availability group listener not empty.\n\n * 400 AgListenerVnetMismatch - Provided VNet for Availability group listener does not match Vnet of internal load balancer.\n\n * 400 InternalLBInvalid - Load balancer specified is not internal.\n\n * 400 InvalidSqlVmResourceIdParameterValue - SQL virtual machine resource id provided is invalid.\n\n * 400 DifferentSubSqlVmList - All SQL virtual machines should be under same subscription.\n\n * 400 OnlyStandardPublicIp - Every virtual machine should have standard public IP.\n\n * 400 ListenerNameTooLong - Listener name should not exceed 15 characters.\n\n * 400 InvalidListenerName - Invalid listener name.\n\n * 400 InvalidLBResourceIdParameterValue - Load balancer resource id is invalid.\n\n * 400 InvalidPublicIpResourceIdParameterValue - Public IP resource id is invalid.\n\n * 400 InvalidSubnetIdParameterValue - Invalid resource id provided for subnet parameter.\n\n * 400 InvalidPrivateIpParameterValue - Invalid address given for private IP address.\n\n * 400 ExtVersionNotSupported - The virtual machine: {0} is running older version of SqlIaasExtension which is not supported by this operation. Please update the extension and retry the operation.\n\n * 400 InvalidReplicaRole - Invalid replica role: {0} specified for SQL VM: {1}.\n\n * 400 InvalidReplicaFailover - Invalid replica Failover: {0} for SQL VM: {1}.\n\n * 400 InvalidReplicaReadableSecondary - Invalid replica readable secondary: {0} for SQL VM: {1}.\n\n * 400 InvalidReplicaCommit - Invalid replica commit mode: {0} for SQL VM: {1}.\n\n * 400 ExpectedSynchronousCommitForAutomaticFailover - Invalid replica commit mode: {0} for SQL VM: {1}. {2} is expected for failover mode: {3}.\n\n * 400 NoPrimaryInAg - There are no replicas with primary role. There should be exactly one replica with role as primary.\n\n * 400 MoreThanOnePrimaryInAg - There are more than one replica with primary role. There should be exactly one replica with role as primary.\n\n * 400 MoreThanThreeSynchronousCommitInAg - There are more than three replica with commit mode as: {0}. There should be at most three replica with commit mode as: {0}\n\n * 400 MoreThanEightSecondariesInAg - There are more than eight secondary replica with secondary role. There should be at most eight replica with role as secondary.\n\n * 400 AlwaysOnNotEnabled - Always on feature is not enabled on Virtual machine: {0}.\n\n * 400 ExpectedReplicaNotPrimary - The VM: {0} does not host the primary replica.\n\n * 400 ListenerSqlvmsMismatchWithReplicaSqlvms - List of SQL VMs specified in Availability group does not match list of SQL VMs in listener config.\n\n * 400 ReplicaSqlvmHasDuplicateEntries - Replica configuration should be unique for each SQL virtual machine. We have found one or more configuration for the same SQL virtual machine resource.\n\n * 400 ReplicaDeleteNotAllowed - Replica deletion through SQL VM RP is not supported yet.\n\n * 400 OnlyBasicPublicIP - All VM specified should contain only Basic Public IP, when providing Loadbalancer Basic Sku.\n\n * 400 NICCannotUseMultipleLBsOfSameType - NIC: {0} cannot reference more than one load balancer of INTERNAL or PUBLIC type\n\n * 400 AGReplicaSQLStartupAccount - NT service account cannot be used as SQL startup account for AG setup. You should use a domain account.\n\n * 400 AgListenerMultiSubnetUpdateNotAllowed - Update of multi subnet AG listener is not allowed.\n\n * 400 ListenerIpSubnetVmSubnetMismatch - Listener static ip address subnet and VM subnet should be same\n\n * 400 NoIpAvailable - No available IP.\n\n * 403 AccessDenied - Access denied.\n\n * 404 ResourceNotFound - The requested resource was not found.\n\n * 404 SubscriptionDoesNotExist - Subscription id does not exist.\n\n * 404 ResourceDoesNotExist - Resource does not exist.\n\n * 404 RPOperationNotFound - Operation not found\n\n * 404 OperationIdNotFound - Operation id could not be found.\n\n * 404 OperationTypeNotFound - Operation Type not found.\n\n * 404 SubscriptionDoesNotHaveSqlVMGroupInResourceGroup - Subscription does not have SQL virtual machine group in resource group.\n\n * 409 OperationInProgress - Operation in progress on resource already.\n\n * 409 OperationCanceled - Operation Cancelled.\n\n * 409 ResourceAlreadyExists - Resource already exists.\n\n * 409 LBGenericErrors - LB operation failed\n\n * 409 NICGenericError - NIC operation failed\n\n * 409 SqlExtensionNotInstalled - SQL extension not installed.\n\n * 409 RPPluginSubstatusMissing - RP plugin substatus missing\n\n * 409 MissingMoveResources - Cannot move resources(s) because some resources are missing in the request.\n\n * 409 ResourceExists - There was an internal error in cleaning up of resources.\n\n * 409 SubscriptionOperationInProgress - An operation on subscription is already in progress\n\n * 409 ReplicasWithDifferentSqlStartupAccounts - Please make sure all your SQL server startup accounts in the specified replicas are the same. This is one of our prerequisites for setting up availability groups.\n\n * 409 ReplicasJoinedToMoreThanOneCluster - AG replicas are joined to more than one failover cluster. Please make sure all your replicas are joined to same failover cluster.\n\n * 409 IPNotAvailable - IP {0} is not available{1}. Consider using one from ({2})\n\n * 409 IpAddressAlreadyReserved - IP address reserved for this listener already exists. Please use the IP address\n\n * 429 TooManyRequestsReceived - Request throttled due to too many requests.\n\n * 429 ArgumentMissingForSetup - Argument is missing for setup. Please provide a value for the argument.\n\n * 429 MultisubnetOperationNotAllowed - Multi subnet HA setup operation is not allowed.\n\n * 500 UnExpectedErrorOccurred - Unexpected error occurred.\n\n * 500 OperationTimeout - Operation timed out.\n\n * 500 RPGenericSystem - System Failure when calling other Resource Provider.\n\n * 500 SqlVmGroupIsBusy - SQL virtual machine group is busy."

add the common error description schema

---
Refers to: specification/sqlvirtualmachine/resource-manager/Microsoft.SqlVirtualMachine/preview/2022-08-01-preview/sqlvm.json:66 in 741ea38. [](commit_id = 741ea38f9997868be9483af427a00827c198423d, deletion_comment = False)

rkmanda commented:
>             "description": "Accepted request to start assessment on SQL virtual machine."

specify x-ms-long-running-operation-options

---
Refers to: specification/sqlvirtualmachine/resource-manager/Microsoft.SqlVirtualMachine/preview/2022-08-01-preview/sqlvm.json:917 in 741ea38. [](commit_id = 741ea38f9997868be9483af427a00827c198423d, deletion_comment = False)

rkmanda commented:
>           ],

fix indent

---
Refers to: specification/sqlvirtualmachine/resource-manager/Microsoft.SqlVirtualMachine/preview/2022-08-01-preview/sqlvm.json:2473 in 741ea38. [](commit_id = 741ea38f9997868be9483af427a00827c198423d, deletion_comment = False)

rkmanda commented:
>             "UnHealthyReplica"

Unhealthy is one word?

---
Refers to: specification/sqlvirtualmachine/resource-manager/Microsoft.SqlVirtualMachine/preview/2022-08-01-preview/sqlvm.json:2472 in 741ea38. [](commit_id = 741ea38f9997868be9483af427a00827c198423d, deletion_comment = False)

rkmanda commented:
>             "description": "*** Error Responses: ***\n\n * 400 InvalidParameterValue - An invalid value was given to parameter.\n\n * 400 RPGenericUser - User Failure when calling other Resource Provider.\n\n * 400 RPResponseMissingAsyncOperationHeader - Response is missing Async operation header information\n\n * 400 InvalidArgument - Invalid argument '{0}'.\n\n * 400 ArgumentNotInRange - Argument '{0}' not in range.\n\n * 400 ResourceNotProvisioned - As the resource: {0} is not in a provisioned state, the request cannot be proceeded forward\n\n * 400 CRPNotAllowedOperation - Operation cannot be completed due to the following error: {0}\n\n * 400 AadAppDisabled - Microsoft AAD app SQLVMResourceProviderAuth ({0}) is disabled in your tenant.\n\n * 400 InternalAPIAccessDenied - Internal API Access denied. App Id {0} is not valid to use this API.\n\n * 400 InvalidSQLAuthPasswordLength - SQL auth password should not be greater than 128 characters\n\n * 400 InvalidRgResourceId - Invalid Resourcegroup resource id specified.\n\n * 400 VmNicVnetMismatch - Virtual machine NIC VNet mismatch.\n\n * 400 NoAvailabilitySet - Vm is not associated with any availability set.\n\n * 400 AvailabilitySetMismatch - Availability set of virtual machines does not match.\n\n * 400 MismatchVmGroupSubscription - Subscription id for SQL virtual machine and SQL virtual machine group are different.\n\n * 400 VmNotRunning - The VM is not in running state.\n\n * 400 VmAgentNotRunning - The VM agent is not installed or in running state.\n\n * 400 AgDoesNotExist - Availability group does not exist.\n\n * 400 AgListDoesNotMatch - SQL Availability group list does not match the list of virtual machines in SQL Virtual machine group.\n\n * 400 NoIpProvided - At least one IP needs to be provided.\n\n * 400 MoreIpProvided - Only one IP needs to be provided.\n\n * 400 SqlVmGroupDoesNotHaveAGListener - SQL virtual machine group does not have the AG listener.\n\n * 400 AgListenerUpdateNotAllowed - Only handful properties of availability group listener can be updated.\n\n * 400 SqlVmNotInGroup - SQL virtual machine is not part of the group.\n\n * 400 SqlVmListCannotBeEmpty - SQL virtual machines list cannot be empty.\n\n * 400 PublicIpNotIPv4 - Public IP must be IPv4 address.\n\n * 400 PublicIPDynamicAllocation - Public IP allocation mode must be static\n\n * 400 PublicLBInvalid - Load balancer specified is not public.\n\n * 400 SubnetMissingFromPrivateIP - Subnet must be provided with private IP.\n\n * 400 IPNotInSubnet - IP not part of subnet.\n\n * 400 NoActivePrimaryInAG - None of the nodes in AvailabilityGroup are Primary\n\n * 400 MultipleListenerSameAG - Multiple availability group listeners for same availability group are not allowed.\n\n * 400 AgListenerNotEmpty - Availability group listener not empty.\n\n * 400 AgListenerVnetMismatch - Provided VNet for Availability group listener does not match Vnet of internal load balancer.\n\n * 400 InternalLBInvalid - Load balancer specified is not internal.\n\n * 400 InvalidSqlVmResourceIdParameterValue - SQL virtual machine resource id provided is invalid.\n\n * 400 DifferentSubSqlVmList - All SQL virtual machines should be under same subscription.\n\n * 400 OnlyStandardPublicIp - Every virtual machine should have standard public IP.\n\n * 400 ListenerNameTooLong - Listener name should not exceed 15 characters.\n\n * 400 InvalidListenerName - Invalid listener name.\n\n * 400 InvalidLBResourceIdParameterValue - Load balancer resource id is invalid.\n\n * 400 InvalidPublicIpResourceIdParameterValue - Public IP resource id is invalid.\n\n * 400 InvalidSubnetIdParameterValue - Invalid resource id provided for subnet parameter.\n\n * 400 InvalidPrivateIpParameterValue - Invalid address given for private IP address.\n\n * 400 ExtVersionNotSupported - The virtual machine: {0} is running older version of SqlIaasExtension which is not supported by this operation. Please update the extension and retry the operation.\n\n * 400 InvalidReplicaRole - Invalid replica role: {0} specified for SQL VM: {1}.\n\n * 400 InvalidReplicaFailover - Invalid replica Failover: {0} for SQL VM: {1}.\n\n * 400 InvalidReplicaReadableSecondary - Invalid replica readable secondary: {0} for SQL VM: {1}.\n\n * 400 InvalidReplicaCommit - Invalid replica commit mode: {0} for SQL VM: {1}.\n\n * 400 ExpectedSynchronousCommitForAutomaticFailover - Invalid replica commit mode: {0} for SQL VM: {1}. {2} is expected for failover mode: {3}.\n\n * 400 NoPrimaryInAg - There are no replicas with primary role. There should be exactly one replica with role as primary.\n\n * 400 MoreThanOnePrimaryInAg - There are more than one replica with primary role. There should be exactly one replica with role as primary.\n\n * 400 MoreThanThreeSynchronousCommitInAg - There are more than three replica with commit mode as: {0}. There should be at most three replica with commit mode as: {0}\n\n * 400 MoreThanEightSecondariesInAg - There are more than eight secondary replica with secondary role. There should be at most eight replica with role as secondary.\n\n * 400 AlwaysOnNotEnabled - Always on feature is not enabled on Virtual machine: {0}.\n\n * 400 ExpectedReplicaNotPrimary - The VM: {0} does not host the primary replica.\n\n * 400 ListenerSqlvmsMismatchWithReplicaSqlvms - List of SQL VMs specified in Availability group does not match list of SQL VMs in listener config.\n\n * 400 ReplicaSqlvmHasDuplicateEntries - Replica configuration should be unique for each SQL virtual machine. We have found one or more configuration for the same SQL virtual machine resource.\n\n * 400 ReplicaDeleteNotAllowed - Replica deletion through SQL VM RP is not supported yet.\n\n * 400 OnlyBasicPublicIP - All VM specified should contain only Basic Public IP, when providing Loadbalancer Basic Sku.\n\n * 400 NICCannotUseMultipleLBsOfSameType - NIC: {0} cannot reference more than one load balancer of INTERNAL or PUBLIC type\n\n * 400 AGReplicaSQLStartupAccount - NT service account cannot be used as SQL startup account for AG setup. You should use a domain account.\n\n * 400 AgListenerMultiSubnetUpdateNotAllowed - Update of multi subnet AG listener is not allowed.\n\n * 400 ListenerIpSubnetVmSubnetMismatch - Listener static ip address subnet and VM subnet should be same\n\n * 400 NoIpAvailable - No available IP.\n\n * 403 AccessDenied - Access denied.\n\n * 404 ResourceNotFound - The requested resource was not found.\n\n * 404 SubscriptionDoesNotExist - Subscription id does not exist.\n\n * 404 ResourceDoesNotExist - Resource does not exist.\n\n * 404 RPOperationNotFound - Operation not found\n\n * 404 OperationIdNotFound - Operation id could not be found.\n\n * 404 OperationTypeNotFound - Operation Type not found.\n\n * 404 SubscriptionDoesNotHaveSqlVMGroupInResourceGroup - Subscription does not have SQL virtual machine group in resource group.\n\n * 409 OperationInProgress - Operation in progress on resource already.\n\n * 409 OperationCanceled - Operation Cancelled.\n\n * 409 ResourceAlreadyExists - Resource already exists.\n\n * 409 LBGenericErrors - LB operation failed\n\n * 409 NICGenericError - NIC operation failed\n\n * 409 SqlExtensionNotInstalled - SQL extension not installed.\n\n * 409 RPPluginSubstatusMissing - RP plugin substatus missing\n\n * 409 MissingMoveResources - Cannot move resources(s) because some resources are missing in the request.\n\n * 409 ResourceExists - There was an internal error in cleaning up of resources.\n\n * 409 SubscriptionOperationInProgress - An operation on subscription is already in progress\n\n * 409 ReplicasWithDifferentSqlStartupAccounts - Please make sure all your SQL server startup accounts in the specified replicas are the same. This is one of our prerequisites for setting up availability groups.\n\n * 409 ReplicasJoinedToMoreThanOneCluster - AG replicas are joined to more than one failover cluster. Please make sure all your replicas are joined to same failover cluster.\n\n * 409 IPNotAvailable - IP {0} is not available{1}. Consider using one from ({2})\n\n * 409 IpAddressAlreadyReserved - IP address reserved for this listener already exists. Please use the IP address\n\n * 429 TooManyRequestsReceived - Request throttled due to too many requests.\n\n * 429 ArgumentMissingForSetup - Argument is missing for setup. Please provide a value for the argument.\n\n * 429 MultisubnetOperationNotAllowed - Multi subnet HA setup operation is not allowed.\n\n * 500 UnExpectedErrorOccurred - Unexpected error occurred.\n\n * 500 OperationTimeout - Operation timed out.\n\n * 500 RPGenericSystem - System Failure when calling other Resource Provider.\n\n * 500 SqlVmGroupIsBusy - SQL virtual machine group is busy."

blocking for this

---
In reply to: [1312323570](https://github.com/Azure/azure-rest-api-specs/pull/21422#issuecomment-1312323570) [](http://example.com/codeflow?ancestors=1312323570,1312323504)

---
Refers to: specification/sqlvirtualmachine/resource-manager/Microsoft.SqlVirtualMachine/preview/2022-08-01-preview/sqlvm.json:66 in f260d7a. [](commit_id = f260d7a64f518bcc513f589c68ea3fa46eeedae1, deletion_comment = False)

qianliu-pdx commented:
> add the common error description schema

Can you point me to some document on how to do it?

qianliu-pdx commented:
> > ```
> >         "description": "Accepted request to start assessment on SQL virtual machine."
> > ```
> 
> specify x-ms-long-running-operation-options
> 
> Refers to: specification/sqlvirtualmachine/resource-manager/Microsoft.SqlVirtualMachine/preview/2022-08-01-preview/sqlvm.json:917 in [741ea38](https://github.com/Azure/azure-rest-api-specs/commit/741ea38f9997868be9483af427a00827c198423d). [](commit_id = [741ea38](https://github.com/Azure/azure-rest-api-specs/commit/741ea38f9997868be9483af427a00827c198423d), deletion_comment = False)

@rkmanda, this specific endpoint was added in previous version when this rule was not enforced. Should we fix the error introduced by the new version or the error from all previous versions?

TimLovellSmith commented:
>             "UnHealthyReplica"

Yes, please fix casing

---
In reply to: [1312326039](https://github.com/Azure/azure-rest-api-specs/pull/21422#issuecomment-1312326039) [](http://example.com/codeflow?ancestors=1312326039,1312326039)

---
Refers to: specification/sqlvirtualmachine/resource-manager/Microsoft.SqlVirtualMachine/preview/2022-08-01-preview/sqlvm.json:2472 in 741ea38. [](commit_id = 741ea38f9997868be9483af427a00827c198423d, deletion_comment = False)

TimLovellSmith commented:
>             "name": "TroubleShootingScenario",

Troubleshooting is actually one word too

---
Refers to: specification/sqlvirtualmachine/resource-manager/Microsoft.SqlVirtualMachine/preview/2022-08-01-preview/sqlvm.json:2476 in 741ea38. [](commit_id = 741ea38f9997868be9483af427a00827c198423d, deletion_comment = False)

TimLovellSmith commented:
>           "description": "Start time in UTC format.",

nit - UTC isn't really a format - its a timezone

---
Refers to: specification/sqlvirtualmachine/resource-manager/Microsoft.SqlVirtualMachine/preview/2022-08-01-preview/sqlvm.json:2461 in 741ea38. [](commit_id = 741ea38f9997868be9483af427a00827c198423d, deletion_comment = False)

TimLovellSmith commented:
>     "SqlVmTroubleshoot": {

A better name might be as SqlVmTroubleshootingDetails?

---
Refers to: specification/sqlvirtualmachine/resource-manager/Microsoft.SqlVirtualMachine/preview/2022-08-01-preview/sqlvm.json:2455 in 741ea38. [](commit_id = 741ea38f9997868be9483af427a00827c198423d, deletion_comment = False)

TimLovellSmith commented:
>     },

Ideally you could do a refactor and remove these definitions and reference common-types everywhere. I seem to remember this is hard for SQL because you have some other way of doing things though?

---
Refers to: specification/sqlvirtualmachine/resource-manager/Microsoft.SqlVirtualMachine/preview/2022-08-01-preview/sqlvm.json:2516 in 741ea38. [](commit_id = 741ea38f9997868be9483af427a00827c198423d, deletion_comment = False)

TimLovellSmith commented:
>             "description": "Accepted request to start SQL virtual machine troubleshoot.",

fix verb case to 'troubleshooting'?

---
Refers to: specification/sqlvirtualmachine/resource-manager/Microsoft.SqlVirtualMachine/preview/2022-08-01-preview/sqlvm.json:1017 in 741ea38. [](commit_id = 741ea38f9997868be9483af427a00827c198423d, deletion_comment = False)

TimLovellSmith commented:
>         "description": "Uninstalls and reinstalls the SQL Iaas Extension.",

Isn't the trend to capitalize it as IaaS?

---
Refers to: specification/sqlvirtualmachine/resource-manager/Microsoft.SqlVirtualMachine/preview/2022-08-01-preview/sqlvm.json:933 in 741ea38. [](commit_id = 741ea38f9997868be9483af427a00827c198423d, deletion_comment = False)

TimLovellSmith commented:
>             "name": "sqlVirtualMachineName",

Is it straightforward for you to refactor this to make it a Parameter that you reuse by `$ref`? Makes reviewing and maintaining slightly easier.

---
Refers to: specification/sqlvirtualmachine/resource-manager/Microsoft.SqlVirtualMachine/preview/2022-08-01-preview/sqlvm.json:940 in 741ea38. [](commit_id = 741ea38f9997868be9483af427a00827c198423d, deletion_comment = False)

TimLovellSmith commented:
>         "description": "Starts Assessment on SQL virtual machine.",

an assessment

---
Refers to: specification/sqlvirtualmachine/resource-manager/Microsoft.SqlVirtualMachine/preview/2022-08-01-preview/sqlvm.json:889 in 741ea38. [](commit_id = 741ea38f9997868be9483af427a00827c198423d, deletion_comment = False)

TimLovellSmith commented:
>         "description": "Starts Assessment on SQL virtual machine.",

what kind of assessment is this?

---
In reply to: [1317435491](https://github.com/Azure/azure-rest-api-specs/pull/21422#issuecomment-1317435491) [](http://example.com/codeflow?ancestors=1317435491)

---
Refers to: specification/sqlvirtualmachine/resource-manager/Microsoft.SqlVirtualMachine/preview/2022-08-01-preview/sqlvm.json:889 in c48a48d. [](commit_id = c48a48d87203041e5afb4a9089256300c60d84d2, deletion_comment = False)

TimLovellSmith commented:
>             "description": "Successfully started assessment on the SQL virtual machine."

What is the response payload schema? Blocking for this

---
Refers to: specification/sqlvirtualmachine/resource-manager/Microsoft.SqlVirtualMachine/preview/2022-08-01-preview/sqlvm.json:911 in 741ea38. [](commit_id = 741ea38f9997868be9483af427a00827c198423d, deletion_comment = False)

TimLovellSmith commented:
>           "200": {

What is the response payload schema? Blocking for this

---
Refers to: specification/sqlvirtualmachine/resource-manager/Microsoft.SqlVirtualMachine/preview/2022-08-01-preview/sqlvm.json:954 in 741ea38. [](commit_id = 741ea38f9997868be9483af427a00827c198423d, deletion_comment = False)

Commit:
addressing one comment from ARM team to add error response schema for
the default response:

Commit:
fix 202 response:
1. adding Location header
2. adding specify x-ms-long-running-operation-options

qianliu-pdx commented:
> UnHealthyReplica

qianliu-pdx commented:
> > ```
> >       ],
> > ```
> 
> fix indent
> 
> Refers to: specification/sqlvirtualmachine/resource-manager/Microsoft.SqlVirtualMachine/preview/2022-08-01-preview/sqlvm.json:2473 in [741ea38](https://github.com/Azure/azure-rest-api-specs/commit/741ea38f9997868be9483af427a00827c198423d). [](commit_id = [741ea38](https://github.com/Azure/azure-rest-api-specs/commit/741ea38f9997868be9483af427a00827c198423d), deletion_comment = False)

The indention seems correct.

Commit:
UnHealthyReplica to UnhealthyReplica
uuHealthyReplica to unhealthyReplica

Commit:
TroubleShoot to Troubleshoot
troubleShoot to troubleshoot

Commit:
"UTC format" to "UTC timezone"

Commit:
SqlVmTroubleshoot to SqlVmTroubleshooting

Commit:
Successfully started SQL virtual machine troubleshoot to Successfully
started SQL virtual machine troubleshooting and similar cases

Commit:
"SQL Iaas Extension" to "SQL IaaS Extension"

Commit:
Assessment to "an assessment" and other similiar issues

Commit:
resolve conflict for custom-words.txt

Commit:
resolve conflict

Commit:
adding default for troubleshootingScenario

Commit:
refactor all sqlVirtualMachineName into common definition and reference
it elsewhere

Commit:
add x-ms-long-running-operation-options with azure-async-operation

Commit:
remove Location header from the Post request

Commit:
add back the Location header for the Post method, as the validation
checks it

TimLovellSmith commented on custom-words.txt:
```diff
@@ -2808,3 +2808,7 @@ routeset
 wayfinding
+dryruns
+Dryrun
+dryrun
```
Are you adding it to custom words because you'rre incorrectly capitalizing these words somewhere? PascalCase should be DryRun, camelCase should be dryRun

TimLovellSmith replied:
I can't actually find you using these words, so you might wnat to just remove them.

qianliu-pdx replied:
This is from resolving the conflicts, If I remove them, I got conflict and the validation does not run

TimLovellSmith commented:
>         "troubleshootStatus": {

I would have probably named it 'troubleshootingStatus'

---
Refers to: specification/sqlvirtualmachine/resource-manager/Microsoft.SqlVirtualMachine/preview/2022-08-01-preview/sqlvm.json:1868 in 8f28bca. [](commit_id = 8f28bca0f4933c331a16336a23440761dba8d74d, deletion_comment = False)

TimLovellSmith commented:
>         "unhealthyReplicaProperties": {

I'm not sure if you really want 'Properties' in the property name?

---
Refers to: specification/sqlvirtualmachine/resource-manager/Microsoft.SqlVirtualMachine/preview/2022-08-01-preview/sqlvm.json:2442 in 8f28bca. [](commit_id = 8f28bca0f4933c331a16336a23440761dba8d74d, deletion_comment = False)

TimLovellSmith commented:
>             "UnhealthyReplica"

Is this eventually going to get used as a discriminator between diferent troubleshooting scenario types?

---
Refers to: specification/sqlvirtualmachine/resource-manager/Microsoft.SqlVirtualMachine/preview/2022-08-01-preview/sqlvm.json:2561 in 8f28bca. [](commit_id = 8f28bca0f4933c331a16336a23440761dba8d74d, deletion_comment = False)

Commit:
add response schema to HA troubleshooting endpoint 200 response.

qianliu-pdx commented:
> > ```
> >         "description": "Successfully started assessment on the SQL virtual machine."
> > ```
> 
> How is someone supposed to get the result of the assessment?
> 
> In reply to: [1317436069](https://github.com/Azure/azure-rest-api-specs/pull/21422#issuecomment-1317436069)
> 
> Refers to: specification/sqlvirtualmachine/resource-manager/Microsoft.SqlVirtualMachine/preview/2022-08-01-preview/sqlvm.json:911 in [741ea38](https://github.com/Azure/azure-rest-api-specs/commit/741ea38f9997868be9483af427a00827c198423d). [](commit_id = [741ea38](https://github.com/Azure/azure-rest-api-specs/commit/741ea38f9997868be9483af427a00827c198423d), deletion_comment = False)

I have talked with the author of this rest endpoint, here is his reply "Start Assessment will trigger an api to actually run the assessment rules. We then collect it and write it to a file that file is picked up by the log analytics workspace and the data is displayed in various ways - power bi report etc ". So there is no need to return anything to the client for the start assessment call except whether it is successful or not.

qianliu-pdx commented:
> UnhealthyReplica

Yes, it will be used as discriminator in the future.

Commit:
add "default" to false for SQL IAAS auto upgrade property

Commit:
renaming:
troubleshootStatus => troubleshootingStatus
unhealthyReplicaProperties => unhealthyReplicaInfo

Commit:
update the HA troubleshooting example after the renaming in this
endpoint

Commit:
rewording Assessment as SQL best practices Assessment after consulting
with PM.

TimLovellSmith commented:
Consider introducing the x-ms-discriminator-value notation to reduce risk of SDK compatibility issues, since that seems to be the way you want things to work in future api versions.

Commit:
revert custom-words.tx file

TimLovellSmith reviewed (approved):
:shipit:

leni-msft commented:
/azp run

qianliu-pdx commented:
Hi Lei,

Some of our rest endpoints require complex preconditions to work. For example, all endpoints related with SQL VM group, SQL HA troubleshooting etc,
Therefore, it cannot be simply tested via the automatic API scenario. Can we have an exemption here?
Also it seems that the Service API readiness test didn't happen in previous API versions. Is it a newly added rule?

Thanks.

Qiang

From: Lei Ni ***@***.***>
Sent: Monday, December 5, 2022 1:03 AM
To: Azure/azure-rest-api-specs ***@***.***>
Cc: Qiang Liu ***@***.***>; Mention ***@***.***>
Subject: Re: [Azure/azure-rest-api-specs] [Hub Generated] Review request for Microsoft.SqlVirtualMachine to add version preview/2022-08-01-preview (PR #21422)


@qianliu-pdx<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fqianliu-pdx&data=05%7C01%7Cqiang.liu%40microsoft.com%7C56c75b7a0c6345555ba708dad69f8814%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638058277921952948%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=X1HWk2W89zi%2BcEQbRjr2JApkQ4%2FKFRcV2r%2BNAwsc5aw%3D&reserved=0> The Service API Readiness Test check<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzure%2Fazure-rest-api-specs%2Fpull%2F21422%2Fchecks%3Fcheck_run_id%3D9881489830&data=05%7C01%7Cqiang.liu%40microsoft.com%7C56c75b7a0c6345555ba708dad69f8814%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638058277921952948%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=T6hl5Y2yazLcRr%2FkxD0PFRfx0dZabEIe2tT1RRBdlYE%3D&reserved=0> failed. The first error seems caused by "InvalidStorageAccountCredentials". Please follow http://aka.ms/ci-fix#service-api-readiness-test<https://nam06.safelinks.protection.outlook.com/?url=http%3A%2F%2Faka.ms%2Fci-fix%23service-api-readiness-test&data=05%7C01%7Cqiang.liu%40microsoft.com%7C56c75b7a0c6345555ba708dad69f8814%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638058277921952948%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=JoebwAkSwcCmf12PPicq2bkRKcWX3RyrFJ3XlgFNcGA%3D&reserved=0> to try fix the examples or define the external dependencies<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzure%2Fazure-rest-api-specs%2Fpull%2F21422%2Fchecks%3Fcheck_run_id%3D9881489830&data=05%7C01%7Cqiang.liu%40microsoft.com%7C56c75b7a0c6345555ba708dad69f8814%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638058277921952948%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=T6hl5Y2yazLcRr%2FkxD0PFRfx0dZabEIe2tT1RRBdlYE%3D&reserved=0> in the baseline scenario.

-
Reply to this email directly, view it on GitHub<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzure%2Fazure-rest-api-specs%2Fpull%2F21422%23issuecomment-1336985715&data=05%7C01%7Cqiang.liu%40microsoft.com%7C56c75b7a0c6345555ba708dad69f8814%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638058277922109223%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=v%2FPGr2iwfVtz98WO4Jz95k7tzBu1vJX3VtDZ1wdTTJU%3D&reserved=0>, or unsubscribe<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAMUHSAES4DAWFVDHV3WRDILWLWVU3ANCNFSM6AAAAAARWYSLXQ&data=05%7C01%7Cqiang.liu%40microsoft.com%7C56c75b7a0c6345555ba708dad69f8814%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638058277922109223%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=Cx4b0Bch5Ss3D00ZdgEbi4BFBUlWIiNdOXLlv7ASmf8%3D&reserved=0>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

akning-ms commented:
Hi @qianliu-pdx , noticed you have cross API version breaking change. did you get breaking change review board sign off?
If no, you may refer to this comment to get sign off. 
https://github.com/Azure/azure-rest-api-specs/pull/21422#issuecomment-1302908698

qianliu-pdx commented:
@akning-ms The breaking changes are due to adding the pattern and length constraint to the  SqlVirtualMachineNameParameter to reject invalid parameters.

The justification are as following:
1) The constraint helps reject invalid parameter earlier at the client and was added upon request from ARM reviewer of this PR
2) It is not in effect a breaking change because 
a) At the backend, our RP will validate this parameter with the same constraint. Therefore the invalid parameter will be caught and client will get an exception anyway. Checking at the client side makes the exception being thrown earlier.
b) This parameter SqlVirtualMachineNameParameter must have the same value of an Azure VM name, which also follows this same pattern. Therefore this pattern does reduce any space of the parameter.
